### PR TITLE
[PIPE-1079] Add automated tests for Spring improvements

### PR DESCRIPTION
**Pull request by Pezzah**
## Goal

Adding more mazerunner tests for the logback appender and spring support

## Review

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/Scenario.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+public abstract class Scenario {
+
+    private static final Logger LOGGER = Logger.getLogger(Scenario.class);
+
+    protected Bugsnag bugsnag;
+
+    public Scenario() {
+        // NOTE: this should already be configured by @Config
+        // all this does is get the instance
+        bugsnag = TestRestController.getBugsnag();
```
Comment (fractalwrench) [inline features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/Scenario.java]: We should pass this in as a constructor parameter instead.

File: features/app_version.feature [hunk 2]
```diff
@@ -7,3 +7,23 @@ Scenario: Test handled Exception
...
+    When I run "LogbackScenario" with logback config "basic_config.xml"
+    Then I should receive a request
+    And the request is a valid for the error reporting API
+    And the "Bugsnag-API-Key" header equals "a35a2a72bd230ac0aa0f52715bbdc6aa"
+    And the payload field "events" is an array with 1 element
+    And the exception "errorClass" equals "java.lang.RuntimeException"
+    And the exception "message" equals "LogbackScenario"
+    And the event "app.version" equals "1.0.0"
+
+Scenario: Test logback appender with app version 1.2.3
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/app_version.feature]: removed one of these

File: features/fixtures/mazerunnerspring/build.gradle [hunk 3]
```diff
@@ -0,0 +1,52 @@
+buildscript {
+    ext {
+        springBootVersion = '1.4.1.RELEASE'
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/fixtures/mazerunnerspring/build.gradle]: Changing this seems to break a lot of things, particularly to do with loggers, I probably need to see what dependencies you are using for this version of spring

File: features/fixtures/mazerunnerspring/build.gradle [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+dependencies {
+    compile("org.springframework.boot:spring-boot-starter")
+    compile("org.springframework.boot:spring-boot-starter-web")
+    compile("ch.qos.logback:logback-classic:1.2.3")
+
+    compile 'com.fasterxml.jackson.core:jackson-annotations:2.9.1'
+    compile 'com.fasterxml.jackson.core:jackson-databind:2.9.1'
+    compile("com.bugsnag:bugsnag:9.9.9-test")
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/fixtures/mazerunnerspring/build.gradle]: It gets built as part of the maze runner setup in `env.rb`

File: features/async_task.feature [hunk 5]
```diff
@@ -0,0 +1,13 @@
+Feature: Reporting Unhandled exceptions in an async task
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/async_task.feature]: renamed

File: features/async_task.feature [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+Scenario: Report an exception from a scheduled task
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/async_task.feature]: added

[the remark this answers is not in the document]
Comment (Pezzah) [reply features/async_task.feature]: This doesn't seem to work very well, I'm not sure how this can be tested properly

File: features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/ScheduledTaskService.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ScheduledTaskService {
+
+    private static final Logger LOGGER = Logger.getLogger(ScheduledTaskService.class);
+
+    private boolean throwException = false;
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/ScheduledTaskService.java]: added `volatile `

File: features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/TestCaseRunner.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+        // Exit the application
+        LOGGER.info("Exiting spring");
+        SpringApplication.exit(ctx, new ExitCodeGenerator() {
+            @Override
+            public int getExitCode() {
+                return 0;
+            }
+        });
+        System.exit(0);
```
[the remark this answers is not in the document]
Comment (Pezzah) [reply features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/TestCaseRunner.java]: changed

File: features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/Scenario.java (hunk 1 above)
Comment (Pezzah) [reply features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/Scenario.java]: Changing this makes things more difficult with constructing the scenarios / passing the parameter to the base class, so I would prefer to leave this as it is

File: features/async_task.feature (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonbowring) [reply features/async_task.feature]: We don't need to test this now as we will recommend that `Future.get()` is wrapped in a try/catch.

File: features/fixtures/mazerunnerspring/build.gradle (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonbowring) [reply features/fixtures/mazerunnerspring/build.gradle]: Changed this to do a `compile project` rather than needing to use a libs dir.

File: features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/Scenario.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            public void close() {
+                // Do nothing
+            }
+        });
+    }
+
+    /**
+     * Flushes sessions from the Bugsnag object
+     */
+    protected void flushAllSessions() {
```
[the remark this answers is not in the document]
Comment (simonbowring) [reply features/fixtures/mazerunnerspring/src/main/java/com/bugsnag/mazerunnerspring/Scenario.java]: It does work 👌
